### PR TITLE
Add DebounceService for iOS

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/33275

Expose the debounce service to iOS. In order to do this a couple of things had to be done:
1. Separate core and common debounce component code so only the usable parts are exposed to iOS
2. Create a debounce service factory for iOS
3. Create a iOS objc API interface for iOS

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/browser/debounce/debounce_service_factory.mm [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        
+  return service;
+}
+
+bool DebounceServiceFactory::ServiceIsNULLWhileTesting() const {
+  return true;
+}
+
+bool DebounceServiceFactory::ServiceIsCreatedWithBrowserState() const {
+  return true;
```
[the remark this answers is not in the document]
Comment (cuba) [reply ios/browser/debounce/debounce_service_factory.mm:74]: I'm not really sure. So yah I'm missing the `debounce_component_installer` call on `brave_application_context_impl` but it doesn't seem to matter. It works just fine. Shouldn't the call to  `debounce::DebounceServiceFactory::GetInstance()` on `brave_browser_state_keyed_service_factories` be enough? 

The other difference on with the two services is that the debounce service takes hold of the component installer, unlike url service which doesn't

[the remark this answers is not in the document]
Comment (kylehickinson) [reply ios/browser/debounce/debounce_service_factory.mm:74]: It works because you are eagerly creating the whole DebounceService which then creates the installer, I'm just pointing out that desktop doesn't do that. They only eagerly set up the installer, the service itself isn't eager.

`GetInstance` is not creating the service, it only sets up the factory

[the remark this answers is not in the document]
Comment (cuba) [reply ios/browser/debounce/debounce_service_factory.mm:70]: Fair enough, I removed this eagerness and eagerly call `debounce_component_installer()`

File: components/brave_shields/adblock/rs/src/convert.rs [hunk 2]
```diff
@@ -12,8 +12,7 @@ use crate::ffi::{
 use adblock::blocker::BlockerResult as InnerBlockerResult;
 use adblock::lists::{ExpiresInterval, FilterListMetadata as InnerFilterListMetadata};
 use adblock::regex_manager::{
-    RegexDebugEntry as InnerRegexDebugEntry,
-    RegexDebugInfo as InnerRegexDebugInfo,
+    RegexDebugEntry as InnerRegexDebugEntry, RegexDebugInfo as InnerRegexDebugInfo,
```
Comment (rillian) [inline components/brave_shields/adblock/rs/src/convert.rs:15]: Are these format changes necessary? Seem unrelated to the rest of the PR.

File: ios/browser/api/de_amp/de_amp_prefs.mm [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/ios/browser/api/de_amp/de_amp_prefs+private.h"
+
+#include "brave/components/de_amp/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "ios/chrome/browser/shared/model/application_context/application_context.h"
+#include "ios/chrome/browser/shared/model/browser_state/chrome_browser_state.h"
+
+@implementation DeAmpPrefs {
```
Comment (bridiver) [inline ios/browser/api/de_amp/de_amp_prefs.mm:13]: shouldn't we just expose a generic interface to prefs? cc @Brandon-T

File: ios/browser/api/debounce/debounce_service+private.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_IOS_BROWSER_API_DEBOUNCE_DEBOUNCE_SERVICE_PRIVATE_H_
+#define BRAVE_IOS_BROWSER_API_DEBOUNCE_DEBOUNCE_SERVICE_PRIVATE_H_
+
+#import <Foundation/Foundation.h>
+
+#include "base/memory/raw_ptr.h"
```
Comment (bridiver) [inline ios/browser/api/debounce/debounce_service+private.h:11]: raw_ptr is not used in this file

File: components/debounce/content/browser/debounce_navigation_throttle.cc [hunk 5]
```diff
@@ -1,9 +1,9 @@
-/* Copyright (c) 2021 The Brave Authors. All rights reserved.
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this file,
- * You can obtain one at https://mozilla.org/MPL/2.0/. */
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Comment (bridiver) [inline components/debounce/content/browser/debounce_navigation_throttle.cc:1]: don't update the copyright year

File: chromium_src/ios/chrome/browser/shared/model/prefs/DEPS [hunk 6]
```diff
@@ -3,6 +3,7 @@ include_rules = [
   "+brave/components/brave_rewards/common",
   "+brave/components/brave_sync",
   "+brave/components/brave_wallet/browser",
+  "+brave/components/de_amp/common",
```
Comment (bridiver) [inline chromium_src/ios/chrome/browser/shared/model/prefs/DEPS:6]: this is fine for now, but we don't need to be this granular with `brave/components`. I think it would be ok to just allow `brave/components` in all of `brave/ios/chrome/browser`

File: ios/browser/debounce/debounce_service_factory.mm [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  return service;
+}
+
+bool DebounceServiceFactory::ServiceIsNULLWhileTesting() const {
+  return true;
+}
+
+web::BrowserState* DebounceServiceFactory::GetBrowserStateToUse(
+    web::BrowserState* context) const {
+  return GetBrowserStateRedirectedInIncognito(context);
```
[the remark this answers is not in the document]
Comment (cuba) [reply ios/browser/debounce/debounce_service_factory.mm:71]: As discussed i'll make a followup PR as other things will need to be moved as well.

File: ios/browser/api/de_amp/de_amp_prefs.mm (hunk 3 above)
Comment (cuba) [reply ios/browser/api/de_amp/de_amp_prefs.mm:13]: I followed along other examples we had but I agree some generic approach would be perhaps nicer.

File: components/brave_shields/adblock/rs/src/convert.rs (hunk 2 above)
Comment (cuba) [reply components/brave_shields/adblock/rs/src/convert.rs:15]: yah it does seem odd. since nothing else touches this file. Ill try to revert this and see what it does.

File: components/brave_shields/adblock/rs/src/engine.rs [hunk 8]
```diff
@@ -6,16 +6,15 @@
...
 use adblock::resources::{MimeType, Resource, ResourceType};
 use adblock::url_parser::ResolvesDomain;
+use adblock::Engine as InnerEngine;
 use cxx::{let_cxx_string, CxxString, CxxVector};
 
 use crate::ffi::{
     resolve_domain_position, BlockerResult, BoxEngineResult, ContentBlockingRulesResult,
-    FilterListMetadata, RegexDebugInfo, RegexManagerDiscardPolicy, UnitResult,
-    VecStringResult,
+    FilterListMetadata, RegexDebugInfo, RegexManagerDiscardPolicy, UnitResult, VecStringResult,
```
Comment (rillian) [inline components/brave_shields/adblock/rs/src/engine.rs:17]: Similar unintended formatting changes here.

File: components/debounce/content/browser/debounce_navigation_throttle.cc (hunk 5 above)
Comment (cuba) [reply components/debounce/content/browser/debounce_navigation_throttle.cc]: Should be fixed

File: ios/browser/api/debounce/debounce_service+private.h (hunk 4 above)
Comment (cuba) [reply ios/browser/api/debounce/debounce_service+private.h]: Fixed

File: ios/browser/debounce/debounce_service_factory.mm [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  std::unique_ptr<debounce::DebounceService> service =
+      std::make_unique<debounce::DebounceService>(
+          braveContext->debounce_component_installer(),
+          braveContext->GetLocalState());
+
+  return service;
+}
+
+bool DebounceServiceFactory::ServiceIsNULLWhileTesting() const {
+  return true;
```
[the remark this answers is not in the document]
Comment (cuba) [reply ios/browser/debounce/debounce_service_factory.mm]: fixed

**Review by kylehickinson — APPROVED** [review]
iOS++

File: ios/browser/api/debounce/debounce_service.mm [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  raw_ptr<debounce::DebounceService> debounceService_;  // NOT OWNED
+}
+
+@end
+
+@implementation DebounceService
+
+- (instancetype)initWithDebounceService:
+    (debounce::DebounceService*)debounceService {
+  self = [super init];
```
Comment (Brandon-T) [inline ios/browser/api/debounce/debounce_service.mm:26]: nit: We should keep this consistent like all other classes.

```objc
if ((self = [super init]]) {

}
```

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: ios/browser/debounce/debounce_service_factory.mm [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+DebounceServiceFactory::DebounceServiceFactory()
+    : BrowserStateKeyedServiceFactory(
+          "DebounceService",
+          BrowserStateDependencyManager::GetInstance()) {}
+
+DebounceServiceFactory::~DebounceServiceFactory() = default;
+
+std::unique_ptr<KeyedService> DebounceServiceFactory::BuildServiceInstanceFor(
+    web::BrowserState* context) const {
+  BraveApplicationContextImpl* braveContext =
```
[the remark this answers is not in the document]
Comment (cuba) [reply ios/browser/debounce/debounce_service_factory.mm:55]: Awesome, yah I saw you merged your PR. This is great.

**Review by Brandon-T — APPROVED** [review]

File: ios/browser/api/debounce/debounce_service.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+OBJC_EXPORT
+@interface DebounceService : NSObject
+- (instancetype)init NS_UNAVAILABLE;
+
+/**
+ True or false if debounce is enabled
+ */
+- (bool)isDebounceEnabled;
```
Comment (kylehickinson) [inline ios/browser/api/debounce/debounce_service.h:20]: This can just be `isEnabled` or `@property(readonly) bool isEnabled;` no need for debounce part

**Review by kylehickinson — APPROVED** [review]

File: components/brave_shields/adblock/rs/src/engine.rs (hunk 8 above)
Comment (cuba) [reply components/brave_shields/adblock/rs/src/engine.rs]: Fixed

File: chromium_src/ios/chrome/browser/shared/model/prefs/DEPS (hunk 6 above)
Comment (cuba) [reply chromium_src/ios/chrome/browser/shared/model/prefs/DEPS]: I updated this file to import brave/components. I hope I understood you correctly.

File: ios/browser/api/debounce/debounce_service.h (hunk 12 above)
Comment (cuba) [reply ios/browser/api/debounce/debounce_service.h]: Changed

Comment (cuba) [reply ios/browser/api/debounce/debounce_service.h]: Actually this was a mistake on my end. I should be exposing the setter too and getting/setting this new value in iOS. though this doesn't really need to be done right away because the default is set to true and the setting is unused within the actual service. But in case this changes, I will include the setter in this PR. Please take a look when you get a chance.

I apologize for the confusion.

File: components/debounce/content/browser/DEPS [hunk 13]
```diff
@@ -0,0 +1,3 @@
+include_rules = [
+  "+content/public/browser",
+]
```
Comment (bridiver) [inline components/debounce/content/browser/DEPS:3]: missing newline here. I recommend configuring your editor to always add it

File: browser/sources.gni [hunk 14]
```diff
@@ -174,6 +174,7 @@ brave_chrome_browser_deps = [
   "//brave/components/cosmetic_filters/common:mojom",
   "//brave/components/de_amp/browser",
   "//brave/components/de_amp/common",
+  "//brave/components/debounce/content/browser",
```
Comment (bridiver) [inline browser/sources.gni:177]: why is this needed here now? I would have expected to see an existing dep for `//brave/components/debounce/browser` if we need this

File: browser/ui/BUILD.gn [hunk 15]
```diff
@@ -548,8 +548,9 @@ source_set("ui") {
     "//brave/components/constants",
     "//brave/components/cosmetic_filters/resources/data:generated_resources",
     "//brave/components/de_amp/common",
-    "//brave/components/debounce/browser",
-    "//brave/components/debounce/common",
+    "//brave/components/debounce/content/browser",
```
Comment (bridiver) [inline browser/ui/BUILD.gn:551]: I don't think we should need `content` here, we only need that for brave_content_browser_client

File: browser/sources.gni (hunk 14 above)
Comment (bridiver) [reply browser/sources.gni:177]: oic, this should go in `brave_browser_debounce_deps` above

Comment (bridiver) [reply browser/sources.gni:177]: oic, brave_browser_debounce_deps handled the deps for the factory which previously included the throttle

File: components/debounce/content/browser/BUILD.gn [hunk 16]
```diff
@@ -1,22 +1,22 @@
...
     "debounce_navigation_throttle.h",
-    "debounce_rule.cc",
-    "debounce_rule.h",
-    "debounce_service.cc",
-    "debounce_service.h",
   ]
 
   deps = [
     "//base",
     "//brave/components/brave_component_updater/browser",
```
Comment (bridiver) [inline components/debounce/content/browser/BUILD.gn:16]: I don't think the component updater dep should be needed here anymore? It also looks like `brave/components/debounce/common/pref_names.h` and `components/user_prefs/user_prefs.h` are unused includes (according to clang) so that could remove some deps as well. Can you double check all the deps here? I see some others like `//brave/extensions:common` that don't appear to be used in the throttle

File: ios/browser/api/de_amp/de_amp_prefs.mm (hunk 3 above)
Comment (bridiver) [reply ios/browser/api/de_amp/de_amp_prefs.mm:13]: can we look at this in a follow-up?

File: ios/browser/api/debounce/debounce_service.mm [hunk 17]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/ios/browser/api/debounce/debounce_service+private.h"
+
+#include <MacTypes.h>
+#include <string>
```
Comment (bridiver) [inline ios/browser/api/debounce/debounce_service.mm:9]: this doesn't appear to be used here

File: ios/browser/api/debounce/debounce_service.mm [hunk 18]
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/ios/browser/api/debounce/debounce_service+private.h"
+
+#include <MacTypes.h>
```
Comment (bridiver) [inline ios/browser/api/debounce/debounce_service.mm:8]: what is this used for?

File: ios/browser/api/debounce/debounce_service.mm (hunk 17 above)
Comment (cuba) [reply ios/browser/api/debounce/debounce_service.mm:9]: nope I removed this. Sorry I sometimes template things out of existing code and remove what is not needed.

File: components/debounce/content/browser/BUILD.gn (hunk 16 above)
Comment (cuba) [reply components/debounce/content/browser/BUILD.gn]: Hmm the CC file I would likely have left it as-is so I probably have some legacy import issues that I didn't address. I removed a bunch of imports there. Thank you for catching these.

But this brings me to a question. Is there a trick to following these imports?

File: ios/browser/api/debounce/debounce_service.mm (hunk 18 above)
Comment (cuba) [reply ios/browser/api/debounce/debounce_service.mm]: To be honest I'm not sure. Same reason as above

**Comment by cuba** [comment]
@bridiver I removed a bunch of unneeded imports. But as stated is there a mechanism in which I can detect unused imports or is it just a manual process of cross referencing imports with code usage?

File: ios/browser/api/debounce/debounce_service.mm (hunk 18 above)
Comment (cuba) [reply ios/browser/api/debounce/debounce_service.mm]: Remove it by the way

**Comment by ShivanKaul** [comment]
> @bridiver I removed a bunch of unneeded imports. But as stated is there a mechanism in which I can detect unused imports or is it just a manual process of cross referencing imports with code usage?

We've discussed adding automation for this but AFAIK we don't have anything just yet.

**Review by bridiver — APPROVED** [review]

File: ios/browser/api/de_amp/de_amp_prefs.mm (hunk 3 above)
Comment (cuba) [reply ios/browser/api/de_amp/de_amp_prefs.mm:13]: Personally I think it's a good idea. But while I opened up some discussion around a generic approach there was no conclusive decision about this.  Im willing to re-open this discussion more openly.
